pytest_datadir/plugin.py

```diff
@@ -25,24 +30,149 @@ def _win32_longpath(path):
-    result = Path(str(tmpdir.join(original_datadir.stem)))
-    if original_datadir.is_dir():
-        shutil.copytree(_win32_longpath(str(original_datadir)), _win32_longpath(str(result)))
-    else:
-        result.mkdir()
-    return result
+
+class DatadirFactory(object):
+    """Factory class for generating datadir fixtures."""
+
+    def __init__(self, request, tmp_path_factory):
+
+        self.tmp_path_factory = tmp_path_factory
+        self.request = request
+
+    def mkdatadir(self, original_datadir='data'):
+
+        # special condition if the datadir is specified as None, which
+        # automatically gets the path that matches the basename of the
+        # module we are in
+        if original_datadir is None:
+            original_datadir = Path(os.path.splitext(self.request.module.__file__)[0])
+
+
+        # get the path to the shared data dir
+        original_path = Path(self.request.fspath.dirname) / original_datadir
+
+        # make sure that the path exists and it is a directory
+        exists = True
+        if not original_path.exists():
+            # raise the flag that it doesn't exist so we can generate
+            # a directory for it instead of copying
+            exists = False
+
+        # make sure the path is a directory if it exists
+        elif not original_path.is_dir():
+            raise ValueError("datadir path is not a directory")
+
+        # generate a base temporary directory and receive the path to it
+        temp_path = self.tmp_path_factory.mktemp('')
+
+        # in order to use the shutil.copytree util the target directory
+        # must not exist so we specify a dir in the generated tempdir for it
+        temp_data_path = temp_path / DATADIR_DIRNAME
+
+        # windows-ify the paths
+        original_path = Path(_win32_longpath(original_path))
+        temp_data_path = Path(_win32_longpath(str(temp_data_path)))
+
+        # copy or create empty directory depending on whether the
+        # original one exists
+        if exists:
+
+            # copy all the files in the original data dir to the temp
+            # dir
+            shutil.copytree(original_path, temp_data_path)
+
+        else:
+            # otherwise just give them a fallback tmpdir
+            temp_data_path.mkdir()
+
+        return temp_data_path
+
+# to clean up for the different scopes we need tmp_path_factory to be
+# scoped differently; just copy the code from module and change the
+# decorator
+
+@pytest.fixture(scope="session")
+def session_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+@pytest.fixture(scope="module")
+def module_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+
+@pytest.fixture(scope="class")
+def class_tmp_path_factory(request):
```

Comment:
I don't think it makes sense to have class/module/function `tmp_path_factory` fixtures if they all return `tmp_path_factory`, which is session-scoped. Session-scoped fixtures can be used from any other scope after all. Also, `DatadirFactory` always copies from a directory named after the module name anyway (`self.request.module.__file__`).

Reply:
In the code for `tmp_path_factory` that this is exactly how it is scoped as session: 
https://github.com/pytest-dev/pytest/blob/0b6960446edce271d0af6dff3293f7eb17f766c1/src/_pytest/tmpdir.py#L155

It returns a `TmpPathFactory` object from the config object which is scope-less and just runs `make_numbered_dir`. So the code here should be enough to change the scope of the fixture.

> Session-scoped fixtures can be used from any other scope after all.

This is the kind of shared state I am trying to avoid :P 

> Also, DatadirFactory always copies from a directory named after the module name anyway (self.request.module.__file__).

Strictly that isn't true, that only happens with the when `original_datadir is None`.

This point though is relevant if your datadir is read-only, in which case you would never need anything other than the session scoped fixture, since you don't really want to be copying multiple times (I am for that being the default behavior). I think there should be the ability to use them read-write and allow for isolation via scopes though.

So basically this is a workaround for the lack of intermediate scopes in `tmp_path_factory` and I am starting to think that this really should be a PR for the main pytest repo for `tmp_path_factory`.

Reply:
Just to make sure we are on the same page, let me ask you a question: when a module-scoped `datadir_factory` goes out of scope (IOW, we start running tests for another module), do you expect that the directories created by `datadir_factory` to be removed?

Reply:
No, I just want a fresh context. So I want `make_numbered_dir` to be called and the basetemp to be set. If I modified the files in the datadir in one module and then need the same datadir for another module the modifications aren't still present.

Reply:
> No, I just want a fresh context.

Ahh I see, OK then. 👍 

Then overall your patch makes sense, I just don't think we need factories for each scope, or even shared-datadirs for each scope too. How about we have only datadir fixtures for each scope?

But perhaps we don't even need the datadir fixtures per-scope variants? I mean:

```python
@pytest.fixture(scope='module')
def mydata(module_datadir):
   store = module_datadir / 'store'
   ...
```

Can just as easily be written as:

```python
@pytest.fixture(scope='module')
def mydata(datadir_factory):
   store = datadir_factory.mkdatadir() / 'store'
   ...
```

Which is exactly the same number of lines of code. Having a small API is beneficial, both from a maintenance POV as well as usability.

@igortg @edisongustavo @tadeu what you guys think?

Reply:
I like the `datadir_factory` approach, looks more simple indeed